### DISCUSSION TITLE
`LinkNeighborLoader`with Hetero graphs and rev edges

### DISCUSSION THREAD

**Discussion by denadai2** [pyg-team/pytorch_geometric#7991] [Q&A] [answered]
Hi pyg,

I am using RandomLinkSplit with rev_edge_types ('book', 'rev_co-read', 'book') and doing a link prediction task with the `LinkNeighborLoader` loader and triplet negative sampling. Does `LinkNeighborLoader` make sure that whenever we sample a positive pair (e.g. "Harry Potter" - co-read - "Lord of The rings") and negative pair, the reverse edge (of the positive pair) is not present in the sampled neighborhood?

thx

**Comment by rusty1s** [comment]
Currently no, `LinkNeighborLoader` will just generate a subgraph from the information present in `edge_index` for the edges present in `edge_label_index`. For these kinds of tasks, you can use `RandomLinkSplit.disjoint_train_ratio` to ensure there is no overlap of edges in `edge_index` and `edge_label_index`. Otherwise, you would have to filter the seed edge (and their reverse) after obtaining the mini-batch output. I agree this is quite a popular request, so it might be good to add something to enable this in PyG right away.

**Comment by denadai2** [comment]
Thanks @rusty1s. I created a small example here, without rev_edges, but with an undirected graph:

```
import torch_geometric
data = torch_geometric.datasets.FakeHeteroDataset(
    num_graphs=1,
    num_node_types=2,
    num_edge_types=2,
    avg_num_nodes=10, avg_degree=2)
print(data.data)
```
Output:
```
HeteroData(
  v0={
    x=[9, 61],
    y=[9],
  },
  v1={ x=[10, 77] },
  (v1, e0, v0)={ edge_index=[2, 16] },
  (v1, e0, v1)={ edge_index=[2, 19] }
)
```

```
from torch_geometric.loader import LinkNeighborLoader

loader = LinkNeighborLoader(
    data.data,
    num_neighbors=[10] * 2,
    batch_size=2,
    edge_label_index=("v1", "e0", "v0"),
    neg_sampling="triplet"
)

sampled_data = next(iter(loader))
print(sampled_data)
```
Output:
```
HeteroData(
  v0={
    x=[4, 61],
    y=[4],
    n_id=[4],
    num_sampled_nodes=[3],
    dst_pos_index=[2],
    dst_neg_index=[2],
  },
  v1={
    x=[10, 77],
    n_id=[10],
    num_sampled_nodes=[3],
    src_index=[2],
  },
  (v1, e0, v0)={
    edge_index=[2, 6],
    e_id=[6],
    num_sampled_edges=[2],
    input_id=[2],
  },
  (v1, e0, v1)={
    edge_index=[2, 16],
    e_id=[16],
    num_sampled_edges=[2],
  }
)
```

```
print(sampled_data["v1"].src_index, sampled_data["v0"].dst_pos_index)
print(sampled_data["v1", "e0", "v0"].edge_index)
```

Output:
```
(tensor([0, 1]), tensor([3, 0]))
tensor([[1, 2, 3, 4, 2, 0],
        [0, 0, 1, 1, 2, 3]])
```

Isn't it strange we try to predict whether 0 is connected to 3, given that 0 IS connected to 3 in `edge_index` and thus it would contribute e.g. in the GNN aggregation?

thanks

**Reply by rusty1s** [reply]
In most cases, this isn't a problem since the model usually has no possibility to remember the existence of a link. This is also inline with what other papers are doing, e.g., LightGCN (as long as validation/test edges are hidden), and may be required if you actually want to do edge classification (where you want to use that edge to make a prediction for it). In general, I agree though that we likely want an option to disable this behavior if it is undesired. Currently, the only option is to separate `edge_label_index` and `edge_index`, or have some custom logic to filter these edges out after mini-batch creation.

**Comment by denadai2** [comment]
thanks, sorry again for the questions but I am kinda new to this. https://snap.stanford.edu/class/cs224w-2020/slides/08-GNN-application.pdf looking at slide 64 I'd have expected they are split. citing Leskovec: "Supervision edges are used as supervision for edge predictions made by the model, will not be fed into GNN!". Thus, I'd specify it in the docs to avoid non-expected behaviours

**Reply by rusty1s** [reply, marked as answer]
Thanks. I clarified this in https://github.com/pyg-team/pytorch_geometric/pull/8832.

**Reply by denadai2** [reply]
thx @rusty1s
